File: Scripts/Python/Barra_carga_constante.py
import numpy as np
import matplotlib.pyplot as plt

# Entrada de dados
L = 1.0  # comprimento total da Barra
E = 100e9  # módulo de elasticidade
A = (10e-3)**2  # área
F = 500  # carga pontual na extremidade direita x=L
po = 2000  # carga distribuída constante
nel = 3  # número de elementos

# Gerando a malha
nnos = nel + 1  # número de nós
he = L / nel  # comprimento do elemento
xn = np.linspace(0, L, nnos)  # coordenadas nodais
inci = np.array([[i, i + 1] for i in range(nnos - 1)])  # formação dos elementos

# Montagem da matriz de Rigidez
Kg = np.zeros((nnos, nnos))
Fg = np.zeros(nnos)

ke = (E * A / he) * np.array([[1, -1], [-1, 1]])  # Rigidez do elemento
fe = (po * he / 2) * np.array([1, 1])  # Carga nodal equivalente > po

for e in range(nel):
    nodes = inci[e, :]
    Kg[np.ix_(nodes, nodes)] += ke
    Fg[nodes] += fe

Fg[-1] += F  # carga Pontual no nó da Direita
freedofs = np.arange(1, nnos)  # aplicar as condições de contorno
uh = np.zeros(nnos)
uh[freedofs] = np.linalg.solve(Kg[np.ix_(freedofs, freedofs)], Fg[freedofs])

# Pós-processamento
NxEl = np.zeros(nel)
for e in range(nel):
    nodes = inci[e, :]
    NxEl[e] = (E * A / he) * np.dot([-1, 1], uh[nodes])

# cálculo da solução analítica
x = np.linspace(0, L, 100)
u = (F / (E * A)) * x + (po / (2 * E * A)) * (2 * L - x) * x
Nx = F + po * (L - x)

# gráficos
plt.figure(figsize=(10, 6))

# Deslocamentos
plt.subplot(2, 1, 1)
plt.plot(xn, uh, 'o-', label='Numérica')
plt.plot(x, u, 'r-x', label='Analítica')
plt.grid(True)
plt.legend()
plt.xlabel('Posição x [m]')
plt.ylabel('Deslocamento u(x) [m]')

# Força Normal
plt.subplot(2, 1, 2)
plt.bar((xn[1:] + xn[:-1]) / 2, NxEl, width=he, align='center', label='Numérica')
plt.plot(x, Nx, 'r-x', label='Analítica')
plt.grid(True)
plt.axis([0, L, -1.5 * abs(np.min(Nx)), 1.5 * abs(np.max(Nx))])
plt.xlabel('Posição x [m]')
plt.ylabel('Carga Normal Nx [N]')
plt.legend()

plt.tight_layout()
plt.show()

File: Scripts/Barra_carga_constante.m
% **************************************************
% Barra carga Constante
%
% E*A*u'' = -po
% u(0) = 0
% du(L)/dx = F/EA
%
% Solução Analítica u(x) = Fx/EA + (po/2EA) * (2L-x) * x
% Força de Reação R(0) = -(poL + F)
%
% Malha de Elementos Finitos
%
% 1-----2-----3-----4 ... node number
% 1 2 3 element number
% **************************************************

close all;
clear all;
clc;

%% Entrada de dados

L = 1; % Comprimento total da Barra
E = 100e9; % Módulo de elasticidade
A = (10e-3)^2; % Área
F = 500; % Carga pontual na extremidade direita x=L
po = 2000; % Carga distribuída constante
nel = 3; % Número de elementos

%% Gerando a malha

nnos = nel + 1; % Número de nós
he = L / nel; % Comprimento do elemento
xn = 0:he:L; % Coordenadas nodais
inci = [[1:nnos-1]' [2:nnos]']; % Formação dos elementos

%% Montagem da matriz de Rigidez

Kg = zeros(nnos);
uh = zeros(nnos, 1);
Fg = zeros(nnos, 1);

ke = (E*A / he) * [1 -1; -1 1]; % Rigidez do elemento
fe = (po * he / 2) * [1; 1]; % Carga nodal equivalente > po

for e = 1:nel
    Kg(inci(e, :), inci(e, :)) = Kg(inci(e, :), inci(e, :)) + ke;
    Fg(inci(e, :), 1) = Fg(inci(e, :), 1) + fe;
end

Fg(end, 1) = Fg(end, 1) + F; % Carga pontual no nó da direita
freedofs = 2:nnos; % Aplicar as condições de contorno
uh(freedofs, 1) = Kg(freedofs, freedofs) \ Fg(freedofs, 1); % Cálculo da resposta

%% Pós-processamento

% cálculo das forças internas
for e = 1:nel
    NxEl(e) = E*A / he * [-1 1] * uh(inci(e, :));
end

%% Cálculo da solução analítica
x = 0:he/5:L;
u = (F / (E*A)) * x + (po / (2 * E * A)) * (2 * L - x) .* x;
Nx = F + po * (L - x);

%% Gráficos
% Deslocamentos
figure(1);
subplot(2, 1, 1);
plot(xn, uh(:, 1), '-o', x, u, 'r-x');
grid on;
legend('Numérica', 'Analítica');
xlabel('Posição x [m]'); ylabel('Deslocamento u(x) [m]');

% Força normal
subplot(2, 1, 2);
bar((xn(2:nel+1) + xn(1:nel)) / 2, NxEl(1:nel), 1);
hold on;
plot(x, Nx, 'r-x');
grid on;
axis([0 L -1.5 * abs(min(Nx)), 1.5 * abs(max(Nx))]);
xlabel('Posição x [m]'); ylabel('Carga Normal Nx [N]');
legend('Numérica', 'Analítica');
hold off;